File: app/android/app/src/main/kotlin/com/friend/ios/PhoneCallsPlugin.kt
package com.friend.ios

import android.content.Context
import android.media.AudioManager
import android.util.Log
import androidx.annotation.NonNull
import com.twilio.voice.Call
import com.twilio.voice.CallException
import com.twilio.voice.ConnectOptions
import com.twilio.voice.Voice
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Flutter plugin for phone call functionality via Twilio Voice SDK.
 * Handles method channel communication between Flutter and native Android.
 * Integrates with Twilio Voice SDK for VoIP calling with real-time audio capture.
 */
class PhoneCallsPlugin private constructor(
    private val context: Context
) : MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

    private var eventSink: EventChannel.EventSink? = null
    private var accessToken: String? = null
    private var activeCall: Call? = null
    private var currentCallId: String? = null
    private var isMuted: Boolean = false
    private var isSpeakerOn: Boolean = false

    companion object {
        private const val TAG = "PhoneCallsPlugin"
        private const val METHOD_CHANNEL = "com.omi/phone_calls"
        private const val EVENT_CHANNEL = "com.omi/phone_calls/events"

        fun registerWith(flutterEngine: FlutterEngine, context: Context) {
            val instance = PhoneCallsPlugin(context)

            val methodChannel = MethodChannel(
                flutterEngine.dartExecutor.binaryMessenger,
                METHOD_CHANNEL
            )
            methodChannel.setMethodCallHandler(instance)

            val eventChannel = EventChannel(
                flutterEngine.dartExecutor.binaryMessenger,
                EVENT_CHANNEL
            )
            eventChannel.setStreamHandler(instance)
        }
    }

    // Twilio Call.Listener for call state callbacks
    private val callListener = object : Call.Listener {
        override fun onRinging(call: Call) {
            Log.d(TAG, "Call ringing")
            sendCallStateEvent("ringing")
        }

        override fun onConnectFailure(call: Call, callException: CallException) {
            Log.e(TAG, "Call failed to connect: ${callException.message}")
            sendCallStateEvent("failed")
            activeCall = null
            currentCallId = null
        }

        override fun onConnected(call: Call) {
            Log.d(TAG, "Call connected")
            activeCall = call
            sendCallStateEvent("active")
        }

        override fun onReconnecting(call: Call, callException: CallException) {
            Log.d(TAG, "Call reconnecting: ${callException.message}")
            sendCallStateEvent("connecting")
        }

        override fun onReconnected(call: Call) {
            Log.d(TAG, "Call reconnected")
            sendCallStateEvent("active")
        }

        override fun onDisconnected(call: Call, callException: CallException?) {
            if (callException != null) {
                Log.e(TAG, "Call disconnected with error: ${callException.message}")
                sendCallStateEvent("failed")
            } else {
                Log.d(TAG, "Call disconnected")
                sendCallStateEvent("ended")
            }
            activeCall = null
            currentCallId = null
        }
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> handleInitialize(call, result)
            "makeCall" -> handleMakeCall(call, result)
            "endCall" -> handleEndCall(result)
            "toggleMute" -> handleToggleMute(call, result)
            "toggleSpeaker" -> handleToggleSpeaker(call, result)
            else -> result.notImplemented()
        }
    }

    // MARK: - Method Handlers

    private fun handleInitialize(call: MethodCall, result: MethodChannel.Result) {
        val token = call.argument<String>("accessToken")
        if (token == null) {
            result.error("INVALID_ARGS", "Missing accessToken", null)
            return
        }

        accessToken = token
        result.success(true)
    }

    private fun handleMakeCall(call: MethodCall, result: MethodChannel.Result) {
        val phoneNumber = call.argument<String>("phoneNumber")
        val callId = call.argument<String>("callId")

        if (phoneNumber == null || callId == null) {
            result.error("INVALID_ARGS", "Missing phoneNumber or callId", null)
            return
        }

        val token = accessToken
        if (token == null) {
            result.error("NOT_INITIALIZED", "Call initialize first", null)
            return
        }

        currentCallId = callId
        sendCallStateEvent("connecting")

        // Connect via Twilio Voice SDK
        val params = HashMap<String, String>()
        params["To"] = phoneNumber
        params["CallId"] = callId

        val connectOptions = ConnectOptions.Builder(token)
            .params(params)
            .build()

        activeCall = Voice.connect(context, connectOptions, callListener)
        result.success(true)
    }

    private fun handleEndCall(result: MethodChannel.Result) {
        activeCall?.disconnect()
        sendCallStateEvent("ended")
        activeCall = null
        currentCallId = null
        result.success(null)
    }

    private fun handleToggleMute(call: MethodCall, result: MethodChannel.Result) {
        val muted = call.argument<Boolean>("muted") ?: false
        isMuted = muted
        activeCall?.mute(muted)
        result.success(null)
    }

    private fun handleToggleSpeaker(call: MethodCall, result: MethodChannel.Result) {
        val speakerOn = call.argument<Boolean>("speakerOn") ?: false
        isSpeakerOn = speakerOn

        val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        audioManager.isSpeakerphoneOn = speakerOn
        result.success(null)
    }

    // MARK: - Event Sending

    private fun sendCallStateEvent(state: String) {
        android.os.Handler(android.os.Looper.getMainLooper()).post {
            eventSink?.success(mapOf("type" to "callStateChanged", "state" to state))
        }
    }

    private fun sendAudioDataEvent(data: ByteArray, channel: Int) {
        android.os.Handler(android.os.Looper.getMainLooper()).post {
            eventSink?.success(mapOf("type" to "audioData", "data" to data, "channel" to channel))
        }
    }

    // MARK: - EventChannel.StreamHandler

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}

File: app/ios/Runner/PhoneCallsPlugin.swift
import Flutter
import UIKit
import AVFoundation
import CallKit
import TwilioVoice

/// Flutter plugin for phone call functionality via Twilio Voice SDK.
/// ConnectOptions.uuid ties Twilio call to CallKit, audio session uses 20ms buffers.
class PhoneCallsPlugin: NSObject, FlutterPlugin {
    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?

    // Twilio Voice
    private var accessToken: String?
    private var activeCall: Call?
    private var callDelegate: TwilioCallDelegateHandler?
    private var currentCallId: String?
    private var isMuted: Bool = false
    private var isSpeakerOn: Bool = false
    private var audioDevice = OmiRecordingAudioDevice()

    // CallKit
    fileprivate let callKitProvider: CXProvider
    private let callKitController = CXCallController()
    fileprivate var callKitUUID: UUID?

    // Pending call info (set before CallKit request, consumed after Twilio connect)
    private var pendingCallResult: FlutterResult?

    override init() {
        let configuration = CXProviderConfiguration()
        configuration.maximumCallGroups = 1
        configuration.maximumCallsPerCallGroup = 1
        configuration.supportsVideo = false
        configuration.supportedHandleTypes = [.phoneNumber]
        callKitProvider = CXProvider(configuration: configuration)
        super.init()
        callKitProvider.setDelegate(self, queue: nil)

        // Wire up audio data callback to stream to Flutter
        audioDevice.onAudioData = { [weak self] data, channel in
            self?.sendAudioDataEvent(data, channel: channel)
        }

        // Set Twilio's audio device (custom device captures both streams)
        TwilioVoiceSDK.audioDevice = audioDevice
    }

    static func register(with registrar: FlutterPluginRegistrar) {
        let instance = PhoneCallsPlugin()

        instance.methodChannel = FlutterMethodChannel(
            name: "com.omi/phone_calls",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: instance.methodChannel!)

        instance.eventChannel = FlutterEventChannel(
            name: "com.omi/phone_calls/events",
            binaryMessenger: registrar.messenger()
        )
        instance.eventChannel?.setStreamHandler(instance)
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            handleInitialize(call, result: result)
        case "makeCall":
            handleMakeCall(call, result: result)
        case "endCall":
            handleEndCall(result: result)
        case "toggleMute":
            handleToggleMute(call, result: result)
        case "toggleSpeaker":
            handleToggleSpeaker(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Method Handlers

    private func handleInitialize(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let token = args["accessToken"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing accessToken", details: nil))
            return
        }

        accessToken = token

        print("PhoneCallsPlugin: initialized with token (length=\(token.count))")
        result(true)
    }

    private func handleMakeCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let phoneNumber = args["phoneNumber"] as? String,
              let callId = args["callId"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing phoneNumber or callId", details: nil))
            return
        }

        guard let token = accessToken else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "Call initialize first", details: nil))
            return
        }

        let contactName = args["contactName"] as? String
        currentCallId = callId

        // Step 1: Generate CallKit UUID
        let uuid = UUID()
        callKitUUID = uuid

        // Step 2: Report outgoing call to CallKit
        let handle = CXHandle(type: .phoneNumber, value: phoneNumber)
        let startCallAction = CXStartCallAction(call: uuid, handle: handle)
        startCallAction.isVideo = false
        startCallAction.contactIdentifier = contactName

        let transaction = CXTransaction(action: startCallAction)

        print("PhoneCallsPlugin: requesting CallKit start for \(phoneNumber)")

        callKitController.request(transaction) { [weak self] error in
            guard let self = self else { return }

            if let error = error {
                print("PhoneCallsPlugin: CallKit start error: \(error.localizedDescription)")
                self.sendCallStateEvent("failed")
                DispatchQueue.main.async { result(false) }
                return
            }

            print("PhoneCallsPlugin: CallKit approved, reporting connecting state")
            self.callKitProvider.reportOutgoingCall(with: uuid, startedConnectingAt: Date())
            self.sendCallStateEvent("connecting")

            // Step 3: Connect via Twilio Voice SDK on main thread
            // pass builder.uuid to tie Twilio call to CallKit call
            DispatchQueue.main.async {
                let delegate = TwilioCallDelegateHandler(plugin: self)
                self.callDelegate = delegate

                let connectOptions = ConnectOptions(accessToken: token) { builder in
                    builder.params = ["To": phoneNumber, "CallId": callId]
                    builder.uuid = uuid  // Ties Twilio call to CallKit UUID
                }

                print("PhoneCallsPlugin: connecting via Twilio SDK...")
                self.activeCall = TwilioVoiceSDK.connect(options: connectOptions, delegate: delegate)

                if self.activeCall != nil {
                    print("PhoneCallsPlugin: Twilio connect returned call object")
                    result(true)
                } else {
                    print("PhoneCallsPlugin: Twilio connect returned nil")
                    self.sendCallStateEvent("failed")
                    result(false)
                }
            }
        }
    }

    private func handleEndCall(result: @escaping FlutterResult) {
        print("PhoneCallsPlugin: ending call")

        // Disconnect the Twilio call
        activeCall?.disconnect()

        // End via CallKit
        if let uuid = callKitUUID {
            let endCallAction = CXEndCallAction(call: uuid)
            let transaction = CXTransaction(action: endCallAction)
            callKitController.request(transaction) { error in
                if let error = error {
                    print("PhoneCallsPlugin: CallKit end error: \(error)")
                }
            }
        }

        sendCallStateEvent("ended")
        cleanup()
        result(nil)
    }

    private func handleToggleMute(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let muted = args["muted"] as? Bool else {
            result(nil)
            return
        }
        isMuted = muted
        activeCall?.isMuted = muted
        audioDevice.isMicStreamMuted = muted
        result(nil)
    }

    private func handleToggleSpeaker(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let speakerOn = args["speakerOn"] as? Bool else {
            result(nil)
            return
        }
        isSpeakerOn = speakerOn

        do {
            let session = AVAudioSession.sharedInstance()
            if speakerOn {
                try session.overrideOutputAudioPort(.speaker)
            } else {
                try session.overrideOutputAudioPort(.none)
            }
        } catch {
            print("PhoneCallsPlugin: speaker toggle error: \(error)")
        }

        result(nil)
    }

    // MARK: - Cleanup

    fileprivate func cleanup() {
        activeCall = nil
        callDelegate = nil
        callKitUUID = nil
        currentCallId = nil
    }

    // MARK: - Event Sending

    func sendCallStateEvent(_ state: String) {
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?(["type": "callStateChanged", "state": state])
        }
    }

    func sendAudioDataEvent(_ data: Data, channel: Int) {
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?([
                "type": "audioData",
                "data": FlutterStandardTypedData(bytes: data),
                "channel": channel,
            ])
        }
    }
}

// MARK: - FlutterStreamHandler

extension PhoneCallsPlugin: FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

// MARK: - Twilio Call Delegate

private class TwilioCallDelegateHandler: NSObject, CallDelegate {
    private weak var plugin: PhoneCallsPlugin?

    init(plugin: PhoneCallsPlugin) {
        self.plugin = plugin
        super.init()
    }

    func callDidStartRinging(call: Call) {
        print("PhoneCallsPlugin: call ringing (sid: \(call.sid))")
        plugin?.sendCallStateEvent("ringing")
    }

    func callDidConnect(call: Call) {
        print("PhoneCallsPlugin: call connected (sid: \(call.sid))")
        // Report connected to CallKit
        if let uuid = plugin?.callKitUUID {
            plugin?.callKitProvider.reportOutgoingCall(with: uuid, connectedAt: Date())
        }
        plugin?.sendCallStateEvent("active")
    }

    func callDidDisconnect(call: Call, error: Error?) {
        if let error = error {
            print("PhoneCallsPlugin: call disconnected with error: \(error.localizedDescription)")
            plugin?.sendCallStateEvent("failed")
        } else {
            print("PhoneCallsPlugin: call disconnected normally")
            plugin?.sendCallStateEvent("ended")
        }

        // Report end to CallKit
        if let uuid = plugin?.callKitUUID {
            let reason: CXCallEndedReason = error != nil ? .failed : .remoteEnded
            plugin?.callKitProvider.reportCall(with: uuid, endedAt: Date(), reason: reason)
        }

        plugin?.cleanup()
    }

    func callDidFailToConnect(call: Call, error: Error) {
        print("PhoneCallsPlugin: call failed to connect: \(error.localizedDescription)")
        plugin?.sendCallStateEvent("failed")

        if let uuid = plugin?.callKitUUID {
            plugin?.callKitProvider.reportCall(with: uuid, endedAt: Date(), reason: .failed)
        }

        plugin?.cleanup()
    }
}

// MARK: - CXProviderDelegate (CallKit)

extension PhoneCallsPlugin: CXProviderDelegate {
    func providerDidReset(_ provider: CXProvider) {
        print("PhoneCallsPlugin: CallKit provider reset")
        activeCall?.disconnect()
        cleanup()
    }

    func provider(_ provider: CXProvider, perform action: CXStartCallAction) {
        print("PhoneCallsPlugin: CallKit CXStartCallAction")
        action.fulfill()
    }

    func provider(_ provider: CXProvider, perform action: CXEndCallAction) {
        print("PhoneCallsPlugin: CallKit CXEndCallAction")
        activeCall?.disconnect()
        sendCallStateEvent("ended")
        action.fulfill()
    }

    func provider(_ provider: CXProvider, perform action: CXSetMutedCallAction) {
        isMuted = action.isMuted
        activeCall?.isMuted = action.isMuted
        audioDevice.isMicStreamMuted = action.isMuted
        action.fulfill()
    }

    func provider(_ provider: CXProvider, didActivate audioSession: AVAudioSession) {
        // 48kHz sample rate, 20ms buffer
        do {
            try audioSession.setCategory(
                .playAndRecord,
                mode: .voiceChat,
                options: [.allowBluetoothHFP, .allowBluetoothA2DP, .defaultToSpeaker]
            )
            try audioSession.setPreferredSampleRate(48000)
            try audioSession.setPreferredIOBufferDuration(0.020)
        } catch {
            print("PhoneCallsPlugin: audio session config error: \(error)")
        }

        print("PhoneCallsPlugin: audio session activated (rate: \(audioSession.sampleRate)Hz, buffer: \(audioSession.ioBufferDuration * 1000)ms)")
        _ = audioDevice.start()
    }

    func provider(_ provider: CXProvider, didDeactivate audioSession: AVAudioSession) {
        print("PhoneCallsPlugin: audio session deactivated")
        _ = audioDevice.stop()
    }
}
